Feature/mega2560

tmclane opened the pull request:
Initial pin / register / port mapping support for Arduino Mega2560.

With these changes `src/examples/blinky1` works perfectly.  I do not have a setup yet to test the rest of the pins since it seems my discrete components (leds, resistors, and the like) are packed away somewhere.  

I used the real pin numbers for all of the pins to be consistent since otherwise the A0-A15 would end up conflicting with D0-D15.  This helped with the math for offsets in some cases.  Unfortunately some of the pin offsets in the masks are semi-random.  I handled those individually where appropriate.

Note:  This is a draft since I do not believe it is complete as of yet.

Addresses #724

tmclane commented:
Compiled and flashed with:


    ./build/tinygo flash -target targets/atmega2560.json examples/blinky1
    ./build/tinygo flash -target targets/atmega2560.json examples/serial

aykevl commented on targets/atmega2560.json:
```diff
@@ -0,0 +1,20 @@
+    ],
+    "ldflags": [
+        "-mmcu=avr6",
+        "-Wl,--defsym=_bootloader_size=0",
```
I looked at the Arduino IDE and it appears that the bootloader is 8kB in size (253952 - 256kB = 8kB):

    mega.menu.cpu.atmega2560.upload.maximum_size=253952

tmclane replied:
> Thank you for this contribution!
> 
> This PR should be rebased on top of the dev branch after #946 has been merged.

Of course this will be done.

tmclane replied:
The comment grouping on Github is being funny. 

I'll make the bootloader change.

aykevl commented on targets/atmega2560.json:
```diff
@@ -0,0 +1,20 @@
+    "inherits": ["avr"],
+    "llvm-target": "avr-atmel-none",
+    "cpu": "atmega2560",
+    "build-tags": ["atmega2560"],
```
Ideally the chip and the board would be separated, but they don't necessarily have to.

* If they are separated, there is one atmega2560.json that defines only the `atmega2560` and `atmega` build tag, and one arduino-mega2560.json that defines board-specific configuration (such as the `arduino_mega2560` build tag, the bootloader size, and the flash command).
* Otherise, this file should be named arduino-mega2560.json and also include the `arduino_mega2560` build tag.

Also, this chip is an ATmega so should define the `atmega` build tag.

tmclane replied:
Ok I was wondering how this should be broken down and what the `target` should be named.

tmclane replied:
I'll see if I can get this split up correctly!

aykevl commented on src/runtime/runtime_atmega.go:
```diff
@@ -1,4 +1,4 @@
-// +build avr,atmega
+// +build avr,atmega avr,atmega2560
```
This change is not necessary with the addition of the `atmega` build tag.

tmclane replied:
> > It might make sense to change our naming scheme a bit. I propose that we change `arduino` -> `arduino-uno`,
> 
> Yes, I think changing the name of the current `arduino` target is a good idea (I've often thought about it but never done it). We could leave an alias for `arduino` but use `arduino-uno` in all examples.
> 
> > and call this target `arduino-mega`.
> 
> Actually, there was an "Arduino Mega" based on the atmega1280. They now call it the "Arduino Mega 2560" to distinguish the two boards. Therefore I think it should be named `arduino-mega2560`.



> > It might make sense to change our naming scheme a bit. I propose that we change `arduino` -> `arduino-uno`,
> 
> Yes, I think changing the name of the current `arduino` target is a good idea (I've often thought about it but never done it). We could leave an alias for `arduino` but use `arduino-uno` in all examples.
> 
> > and call this target `arduino-mega`.
> 
> Actually, there was an "Arduino Mega" based on the atmega1280. They now call it the "Arduino Mega 2560" to distinguish the two boards. Therefore I think it should be named `arduino-mega2560`.

Should this type of renaming be done in this same PR?
I'll make them separate commits at the very least.

aykevl replied:
> > Actually, there was an "Arduino Mega" based on the atmega1280. They now call it the "Arduino Mega 2560" to distinguish the two boards. Therefore I think it should be named `arduino-mega2560`.
> 
> Should this type of renaming be done in this same PR?
> I'll make them separate commits at the very least.

The Arduino Mega, yes. But not `arduino` -> `arduino-uno`, that's a separate matter.

aykevl commented on src/machine/machine_atmega328p.go:
```diff
@@ -0,0 +1,15 @@
+const irq_USART0_RX = avr.IRQ_USART_RX
+
+// LED on the Arduino
+const LED Pin = 13
```
The LED pin is board specific so should be defined in board_arduino-mega2560.go (with the appropriate build tag: `// +build arduino_mega2560`).

tmclane replied:
Ok thanks for the clarification.  It was unclear to me what pieces went in which files.  I have seen mixtures in other files that contradicted themselves.

aykevl commented on src/machine/board_arduino.go:
```diff
@@ -1,15 +1,7 @@
-// +build arduino
+// +build arduino,!atmega2560
```
This change should not be necessary.

aykevl commented on src/machine/board_arduino.go:
```diff
@@ -1,15 +1,7 @@
-
-// LED on the Arduino
-const LED Pin = 13
-
```
This appears to be leftover from testing other changes?

tmclane replied:
believe it is actually.  I was trying to fix the CircleCI wasm build.

jaddr2line commented:
It might make sense to change our naming scheme a bit. I propose that we change `arduino` -> `arduino-uno`, and call this target `arduino-mega`.

aykevl commented on src/machine/board_arduino_mega2560.go:
```diff
@@ -0,0 +1,88 @@
+// +build atmega2560
```
This build tag should reflect the file name: `arduino_mega2560`.

aykevl commented:
> It might make sense to change our naming scheme a bit. I propose that we change `arduino` -> `arduino-uno`,

Yes, I think changing the name of the current `arduino` target is a good idea (I've often thought about it but never done it). We could leave an alias for `arduino` but use `arduino-uno` in all examples.

>  and call this target `arduino-mega`.

Actually, there was an "Arduino Mega" based on the atmega1280. They now call it the "Arduino Mega 2560" to distinguish the two boards. Therefore I think it should be named `arduino-mega2560`.

tmclane commented:
I have :

- rebased my branch back on top of the `atmega1284` branch
- refactored the targets
  - targets/atmega2560.json - atmega2560 generic settings
  - targets/arduino-mega2560.json - Arduino Mega2560 specific settings
- refactored the board / platform code
  - src/machine_atmega2560.go - machine specific settings (currently only the const definition)
  - src/board_arduino_mega2560.go - Arduino Mega 2560 pin definitions + pin function definitions

**Note:**  I had to add a build exclude tag to the `machine_atmega.go` file since otherwise the board specific functions like `func (p Pin) Configure(config PinConfig)` were defined twice.
I believe that this would be solved by moving the code out of `machine_atmega.go` and into a `Arduino Uno` specific file (which wasn't intended to be included in this PR).

aykevl commented:
> **Note:** I had to add a build exclude tag to the `machine_atmega.go` file since otherwise the board specific functions like `func (p Pin) Configure(config PinConfig)` were defined twice.

There is a difference between board files (like `board_arduino_mega2560.go`) and chip files (like `machine_atmega2560.go`).
This exclude really needs to be removed. Where that leads to a conflict because a function is defined multiple times:
* if it is the same for all the atmega chips (atmega328p, atmega2560, etc), it should be in the shared `machine_atmega.go` file.
* If it needs to be different, move it to more specific files.

For example, it seems like UART configuration can be shared between chips so `(UART) Configure` can remain in machine_atmega.go while a function like `getPortMask` will need to be moved to more specific files (`machine_atmega328p.go` and `machine_atmega2560.go`).

tmclane commented:
Ok I tested all digital pins 0 - 53 and found a few bugs and fixed them.
I am going to check the Analog pins next and figure out why PWM doesn't work as well.

tmclane commented:
Status update:

- A0 & A1 will not output anything in digital mode
- ADC works except that A0 and A1 float high unless grounded

deadprogram commented on src/machine/board_arduino_mega2560.go:
```diff
@@ -0,0 +1,199 @@
+
+// getRegisterPortMask returns the register, port, and mask for the pin
+func (p Pin) getRegisterPortMask() (*volatile.Register8, *volatile.Register8, uint8) {
+	if p <= A0 && p > A8 { // A0-A7
```
It would be a lot better to use a `switch` statement here. Also, it appears that the condition here is the opposite to what the comment says it is?

tmclane replied:
I kept it as separate `if` statements because the original I copied as a guide used them (`machine_atmega.go` - before this branch)

Why do you think the comment is wrong?
it's not a subtraction operator in this case:  

```	A0  Pin = 97 // PF0
	A1  Pin = 96 // PF1
	A2  Pin = 95 // PF2
	A3  Pin = 94 // PF3
	A4  Pin = 93 // PF4
	A5  Pin = 92 // PF5
	A6  Pin = 91 // PF6
	A7  Pin = 90 // PF7
	A8  Pin = 89 // PK0
```

I guess I should have used set notation?

tmclane replied:
The comments were intended to be pin ranges not mathematical operations.

tmclane replied:
I'll update the code to use a `switch` statement.  I had thought about that originally but thought there was some reason not to since I hadn't seen one used anywhere (although I didn't look very hard)

deadprogram replied:
My bad, I only just noticed that the A0 thru A9 is in descending order. Most of the boards this is ascending.

Either way, my comment about `switch` instead of a whole bunch of `if` statements still stands, if you please.

tmclane commented:
@deadprogram if => switch has been done. 

 For the `default` case what do you recommend to return? 
I am not happy with what is being returned at the moment.  I suppose the callers of this function could check for a bad mask and silently discard the request?

deadprogram commented:
One option for the default case is to panic, which should happen at compile time causing the compile to fail.

deadprogram commented:
BTW just tested blinky1 on both my arduino uno and also my mega2560 board, and it worked as expected on both.

tmclane commented on src/machine/board_arduino_mega2560.go:
```diff
@@ -0,0 +1,174 @@
+// getRegisterPortMask returns the register, port, and mask for the pin
+func (p Pin) getRegisterPortMask() (*volatile.Register8, *volatile.Register8, uint8) {
+	switch p {
+	case A0, A1, A2, A3, A4, A5, A6, A7:
```
Ok.  That makes sense.

tmclane replied:
If I use `PF0` when I define it is it ok to use the ATMega2560 pin number for it's value?

i.e. 

```go
const (
    PF0 = 97  // A0 (on Mega 2560)

)
```

Or do you recommend just an `iota` sequence?

aykevl replied:
Yes that's the intention. But you could perhaps make it more readable with something like this:

```go
const (
    PF0 Pin = 5*8 + 0
    PF1 Pin = 5*8 + 1
    PF2 Pin = 5*8 + 2
)
```
You could also define `portA`, `portB` constants like [here](https://github.com/tinygo-org/tinygo/blob/master/src/machine/board_stm32.go#L7) and write:

```go
const (
    PF0 = portF + 0
    PF1 = portF + 1
    PF2 = portF + 2
)
```


Where did you get 97 from? Or was it just an example?

tmclane replied:
97 is the Atmega2560 pin number according to my diagram of an Arduino Mega2560.

tmclane replied:
Just checked the datasheet:  97 == A0

http://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-2549-8-bit-AVR-Microcontroller-ATmega640-1280-1281-2560-2561_datasheet.pdf

tmclane replied:
so it isn't board specific.  (was slightly worried it could be)

tmclane commented on src/machine/board_arduino_mega2560.go:
```diff
@@ -0,0 +1,174 @@
+	ADC15 Pin = 15
+
+	// Digital pins
+	D0  Pin = 2  // PE0
```
I will shift to this style.

tmclane commented:
@aykevl Hopefully this latest push is more in line with what you were expecting.

tmclane commented:
FYI: I am thinking about changing the subtraction targets to be the `port<X>` constants instead of the lowest pin on each port.  Would the be nicer looking?

aykevl commented on targets/arduino-mega2560.json:
```diff
@@ -0,0 +1,8 @@
+    "inherits": ["atmega2560"],
+    "build-tags": ["arduino_mega2560"],
+    "ldflags": [
+        "-Wl,--defsym=_bootloader_size=253952"
```
It's the other way round: 253952 is the size available to the program (according to boards.txt in the Arduino IDE). The chip has 256kB of memory (262144) so the bootloader size must be 8192, or 8K.

```suggestion
        "-Wl,--defsym=_bootloader_size=8192"
```

tmclane replied:
Ok I made this change.  I guess I misunderstood an earlier conversation when I put that in place.

tmclane replied:
Where do I place the `253952` instead?  Is there another `defsym` value?

aykevl replied:
Nowhere, it's just a different way of indicating which area of flash is available for the program.

aykevl commented:
Thank you! I've now merged this PR into the dev branch so it'll be part of the next release.

tmclane commented:
awesome to hear!
